File: java/com/google/copybara/git/gitlab/GitLabUtil.java
/*
 * Copyright (C) 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.copybara.git.gitlab;

import static java.nio.charset.StandardCharsets.UTF_8;

import java.net.URI;
import java.net.URLEncoder;
import java.util.Locale;

/** Utility class for GitLab endpoints. */
public class GitLabUtil {
  private GitLabUtil() {}

  /**
   * Returns the URL-encoded project path for a given GitLab repository URL.
   *
   * <p>The project path is defined as the namespace plus the project name.
   *
   * <p>The URL encoded project path is used for querying merge requests, and possibly other
   * entities, from the public REST API.
   *
   * @param repoUrl the URL to extract the project path from
   * @return the encoded project path
   * @see <a href="https://docs.gitlab.com/api/rest/#namespaced-paths">GitLAb API Namespaced paths
   *     docs</a>
   */
  public static String getUrlEncodedProjectPath(URI repoUrl) {
    String path = repoUrl.getPath().trim().toLowerCase(Locale.ROOT);
    // Remove any leading '/'
    while (path.startsWith("/")) {
      path = path.substring(1);
    }
    // Remove any trailing .git
    if (path.endsWith(".git")) {
      path = path.substring(0, path.length() - ".git".length());
    }
    return URLEncoder.encode(path, UTF_8);
  }
}

File: src/Copybara.Core/Git/GitLab/GitLabUtil.cs
/*
 * Copyright (C) 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Text;

namespace Copybara.Git.GitLab;

/// <summary>
/// Utility class for GitLab endpoints. Port of
/// <c>com.google.copybara.git.gitlab.GitLabUtil</c>.
/// </summary>
public static class GitLabUtil
{
    /// <summary>
    /// Returns the URL-encoded project path for a given GitLab repository URL.
    ///
    /// <para>The project path is defined as the namespace plus the project name.</para>
    ///
    /// <para>The URL encoded project path is used for querying merge requests, and possibly other
    /// entities, from the public REST API.</para>
    /// </summary>
    /// <param name="repoUrl">the URL to extract the project path from.</param>
    /// <returns>the encoded project path.</returns>
    public static string GetUrlEncodedProjectPath(Uri repoUrl)
    {
        string path = repoUrl.AbsolutePath.Trim().ToLowerInvariant();
        // Remove any leading '/'.
        while (path.StartsWith('/'))
        {
            path = path.Substring(1);
        }
        // Remove any trailing .git.
        if (path.EndsWith(".git", StringComparison.Ordinal))
        {
            path = path.Substring(0, path.Length - ".git".Length);
        }
        return UrlEncode(path);
    }

    // Mirrors java.net.URLEncoder.encode(path, UTF_8) (application/x-www-form-urlencoded).
    private static string UrlEncode(string value)
    {
        var sb = new StringBuilder();
        foreach (byte b in Encoding.UTF8.GetBytes(value))
        {
            char c = (char)b;
            if ((c >= 'A' && c <= 'Z')
                || (c >= 'a' && c <= 'z')
                || (c >= '0' && c <= '9')
                || c is '-' or '_' or '.' or '*')
            {
                sb.Append(c);
            }
            else if (c == ' ')
            {
                sb.Append('+');
            }
            else
            {
                sb.Append('%');
                sb.Append(((int)b).ToString("X2"));
            }
        }
        return sb.ToString();
    }
}
